kafka-persistence/src/main/scala/com/evolutiongaming/kafka/flow/kafkapersistence/KafkaPersistence.scala

```diff
@@ -0,0 +1,166 @@
+    }
+
+    apply(
+      keysOf = keysOf,
+      snapshots = PersistenceOf.snapshotsOnly(keysOf, snapshotsOf),
+      onRecoveryFinished = MonadState.set(BytesByKey.zero)
+    )
+  }
+
+  private def readState[F[_] : BracketThrowable : FromBytes[*[_], String] : Log](
+                                                                                  consumerOf: ConsumerOf[F],
+                                                                                  consumerConfig: ConsumerConfig,
+                                                                                  stateTopic: Topic,
+                                                                                  partition: Partition
+                                                                                ): F[BytesByKey] = {
+    consumerOf
+      .apply[String, ByteVector](
+        consumerConfig.copy(common = consumerConfig.common.copy(clientId = consumerConfig.common.clientId.map(cid => s"$cid-state-$partition")))
+      )
+      .use { consumer =>
+        val statePartition = TopicPartition(topic = stateTopic, partition = partition)
+
+        def processRecord(map: BytesByKey, record: ConsRecord): BytesByKey = record match {
+          case ConsumerRecord(_, _, _, Some(WithSize(key, _)), Some(WithSize(value, _)), _) =>
+            map + (key -> value)
+          case ConsumerRecord(_, _, _, Some(WithSize(key, _)), None, _) => map - key
+          case _ => map //ignore records with no key for now
+        }
+
+        def readPartition(targetOffset: Offset): F[BytesByKey] =
+          Log[F].info(s"State topic read started up to offset $targetOffset") *>
+            FlatMap[F]
+              .tailRecM[BytesByKey, BytesByKey](BytesByKey.zero) { acc =>
+                consumer
+                  .position(statePartition)
+                  .flatMap {
+                    case offset if offset >= targetOffset =>
+                      acc.asRight[BytesByKey].pure[F]
+                    case _ =>
+                      consumer
+                        .poll(10.millis)
+                        .map(_.values.values.flatten.foldLeft(acc)(processRecord).asLeft)
+                  }
+              }
+              .flatTap { map =>
+                Log[F].info(
+                  s"State topic $stateTopic partition $partition read complete at offset $targetOffset, ${map.size} keys read"
+                )
+              }
+
+        val statePartitionSingleton = data.NonEmptySet.of(statePartition)
+        for {
+          _ <- consumer.assign(statePartitionSingleton)
+          endOffsets <- consumer.endOffsets(statePartitionSingleton)
+          targetOffset <- BracketThrowable[F].fromOption(
+            endOffsets.get(statePartition),
+            MissingOffsetError(statePartition)
+          )
+          bytesByKey <- readPartition(targetOffset)
+        } yield bytesByKey
+      }
+  }
+
+  implicit def nelAsIterable[A]: NonEmptyList[A] => IterableOnce[A] = _.toList
```

Comment:
Just call `toList` explicitly where it is needed?

Reply:
Done (with `toIterable`)